File: PHP/DesignPattern/FactoryMethod/PizzaStore.php
<?php
	/**
	* 抽象创建者类
	* 抽象产品由子类制造，创建者不需要真的知道在制造哪种具体产品
	*/
	abstract class PizzaStore {
		public function orderPizza($type) {
			$pizza = $this->createPizza($type);
			
			$pizza->prepare();
			$pizza->bake();
			$pizza->cut();
			$pizza->box();
			return $pizza;			
		}

		/**
		* 一个抽象的工厂方法，让子类实现此方法制造产品
		*/
		protected abstract function createPizza($type);
	}

File: Java/FactoryMethod/src/com/base/PizzaStore.java
package com.base;

import com.base.Pizza;

/**
 * Created by hhq on 1/14/15.
 */
public abstract class PizzaStore {
    public Pizza orderPizza(String type) {
        Pizza pizza;
        
        pizza = createPizza(type);
        
        pizza.prepare();
        pizza.bake();
        pizza.cut();
        pizza.box();
        return pizza;
    }
    
    protected abstract Pizza createPizza(String type);
}
